### PR TITLE
Enable Dora write through worker to UFS

### PR DESCRIPTION
**Pull request by huanghua78**
### What changes are proposed in this pull request?

- Clients write data to worker and worker write data to local paging cache and UFS. 
   Added a switch to control this. The default way is client writing directly to UFS.
- WriteType.THROUGH is handled.
- InAlluxioPercentage and inMemoryPercentage is set to 100 when file length is zero.

### Why are the changes needed?

So Alluxio workers can do some control, e.g. security, flow, quota, on worker. This also avoid managing UFS access tokens on clients, minimize performance bottles on client.
TODO: worker decides writing data to local paging cache or remote UFS based on different write types.
For example, for CACHE_THROUGH, data is going to local paging store and UFS.
For THROUGH, data is ONLY going to UFS.
### Does this PR introduce any user facing changes?

N/A

### CODE REVIEW

**Comment by JiamingMai** [comment]
Could you add some tests for this change? Otherwise, it would be more and more difficult to check whether the behavior of DORA worker is correct or not.

**Comment by huanghua78** [comment]
> Could you add some tests for this change? Otherwise, it would be more and more difficult to check whether the behavior of DORA worker is correct or not.


Existing System test and Integration test, Unit test have already covered write APIs from client.

File: dora/core/client/fs/src/main/java/alluxio/client/file/DoraCacheFileSystem.java [hunk 1]
```diff
@@ -275,10 +275,8 @@ public FileOutStream createFile(AlluxioURI alluxioPath, CreateFilePOptions optio
       outStreamOptions.setMountId(status.getMountId());
       outStreamOptions.setAcl(status.getAcl());
 
-      FileOutStream ufsOutStream = mDelegatedFileSystem.createFile(ufsFullPath, options);
-
       FileOutStream doraOutStream = mDoraClient.getOutStream(ufsFullPath, mFsContext,
-          outStreamOptions, ufsOutStream, uuid);
```
Comment (yyongycy) [inline dora/core/client/fs/src/main/java/alluxio/client/file/DoraCacheFileSystem.java:281]: is ufsOutStream still supported? Not find the replacing logic.

File: dora/core/client/fs/src/main/java/alluxio/client/file/DoraCacheFileSystem.java [hunk 2]
```diff
@@ -275,10 +275,8 @@ public FileOutStream createFile(AlluxioURI alluxioPath, CreateFilePOptions optio
       outStreamOptions.setMountId(status.getMountId());
       outStreamOptions.setAcl(status.getAcl());
 
-      FileOutStream ufsOutStream = mDelegatedFileSystem.createFile(ufsFullPath, options);
```
Comment (beinan) [inline dora/core/client/fs/src/main/java/alluxio/client/file/DoraCacheFileSystem.java:278]: Is it possible to keep this on client? we can have a flag to switch on and off write through worker

Comment (huanghua78) [reply dora/core/client/fs/src/main/java/alluxio/client/file/DoraCacheFileSystem.java:278]: Yes. Please see Line 283 - 288.
There is still a fallback to UFS, and there is a switch to enable/disable such fallback.

File: dora/core/client/fs/src/main/java/alluxio/client/file/DoraCacheFileSystem.java (hunk 1 above)
Comment (huanghua78) [reply dora/core/client/fs/src/main/java/alluxio/client/file/DoraCacheFileSystem.java:281]: Yes. Line 283 - 288 is fallback to UFS from client.
The same logic as all other operations.

There is also a switch to control the fallback.

Comment (yyongycy) [reply dora/core/client/fs/src/main/java/alluxio/client/file/DoraCacheFileSystem.java:281]: Got it.

**Comment by huanghua78** [comment]
Mark it as draft, since it needs more discussion on this feature.

**Comment by yyongycy** [comment]
I think it should be fine to have this by default disable

**Review by yyongycy — APPROVED** [review]
lgtm

**Comment by huanghua78** [comment]
@yyongycy Please take another look. Handled a corner case:
For THOUGH, and client_write_to_ufs disabled.
In this case, client should send data to worker.

**Review by yyongycy — APPROVED** [review]
lgtm

**Comment by huanghua78** [comment]
@beinan please review this PR.

File: dora/core/client/fs/src/main/java/alluxio/client/file/DoraFileOutStream.java [hunk 3]
```diff
@@ -151,17 +156,23 @@ public void close() throws IOException {
       if (mUnderStorageType.isSyncPersist()) {
         try {
           if (mCanceled) {
-            mUnderStorageOutputStream.cancel();
+            if (mUnderStorageOutputStream != null) {
```
Comment (elega) [inline dora/core/client/fs/src/main/java/alluxio/client/file/DoraFileOutStream.java:159]: nit:
if (mUnderStorageOutputStream != null) {
    if (mCancelled) {
mUnderStorageOutputStream.cancel();
   } else {
mUnderStorageOutputStream.flush();
}
}

**Review by elega — COMMENTED** [review]
LGTM
can we also add a unit/integration test to cover the write through worker case.
I think DoraFileSystemIntegrationTest is a good place to test cover it.

File: dora/core/common/src/main/java/alluxio/worker/dora/OpenFileHandle.java [hunk 4]
```diff
@@ -30,13 +31,17 @@ public class OpenFileHandle {
   private OutputStream   mOutStream; //outstream from UFS
```
Comment (JiamingMai) [inline dora/core/common/src/main/java/alluxio/worker/dora/OpenFileHandle.java:31]: Can we rename this field to `mUfsOutStream `for more readable?

**Review by JiamingMai — APPROVED** [review]
LGTM

File: dora/core/client/fs/src/main/java/alluxio/client/file/DoraFileOutStream.java (hunk 3 above)
Comment (huanghua78) [reply dora/core/client/fs/src/main/java/alluxio/client/file/DoraFileOutStream.java:159]: OK

File: dora/core/common/src/main/java/alluxio/worker/dora/OpenFileHandle.java (hunk 4 above)
Comment (huanghua78) [reply dora/core/common/src/main/java/alluxio/worker/dora/OpenFileHandle.java:31]: OK

**Comment by huanghua78** [comment]
> LGTM
> can we also add a unit/integration test to cover the write through worker case.
> I think DoraFileSystemIntegrationTest is a good place to test cover it.

UT is modified to cover this new switch property.

**Comment by huanghua78** [comment]
@elega please review again.

**Review by elega — APPROVED** [review]

**Comment by huanghua78** [comment]
alluxio-bot, merge this please

**Pull request merged**